client/src/jobs/job.py

```diff
@@ -389,36 +267,72 @@ def __init__(
         if (module := inspect.getmodule(self._func)) is None:
             raise ValueError("Cannot derive module for Job function.")
 
+        self._file = Path(str(module.__file__))
         self._name = self._func.__name__
-        self._file = os.path.relpath(str(module.__file__))
-
+        self.build_context = (
+            build_context
+            if build_context is not None
+            else self._resolve_build_context(self._file)
+        )
         self.validate()
 
+    def _resolve_build_context(self, job_file: Path) -> Path:
+        build_context = job_file
+        max_depth = 15
+        for _ in range(max_depth):
+            if self._is_project_root(build_context):
+                break
+            build_context = build_context.parent
+        else:
+            raise ValueError(
+                f"Could not resolve build context from job file {self._file}, traversed {max_depth} up."
+            )
+        return build_context
+
+    @classmethod
+    def _is_project_root(cls, path: Path) -> bool:
+        indicators = [
+            ".git",
+            "pyproject.toml",
+            "setup.py",
+        ]
+        return any((path / indicator).exists() for indicator in indicators)
+
     @property
     def name(self) -> str:
         return self._name
 
     @property
     def file(self) -> str:
-        return self._file
+        return str(self._file)
 
     def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T:
         return self._func(*args, **kwargs)
 
+    def _resolve_path_in_build_context(self, entrypoint: Path) -> Path:
```

Comment:
`entrypoint` is already an existing term in the container world, I'd go for just `path` instead.